source/common/network/udp_listener_impl.cc

```diff
@@ -34,20 +35,52 @@ UdpListenerImpl::UdpListenerImpl(Event::Dispatcher& dispatcher, SocketSharedPtr
     : BaseListenerImpl(dispatcher, std::move(socket)), cb_(cb), time_source_(time_source),
       // Default prefer_gro to false for downstream server traffic.
       config_(config, false) {
+  parent_drained_callback_registrar_ = socket_->parentDrainedCallbackRegistrar();
   socket_->ioHandle().initializeFileEvent(
       dispatcher, [this](uint32_t events) -> void { onSocketEvent(events); },
-      Event::PlatformDefaultTriggerType, Event::FileReadyType::Read | Event::FileReadyType::Write);
+      Event::PlatformDefaultTriggerType, paused() ? 0 : events_when_unpaused_);
+  if (paused()) {
+    parent_drained_callback_registrar_->registerParentDrainedCallback(
+        socket_->connectionInfoProvider().localAddress(),
+        [this, &dispatcher, alive = std::weak_ptr<void>(destruction_checker_)]() {
+          dispatcher.post([this, alive = std::move(alive)]() {
+            if (alive.lock()) {
```

Comment:
I think to be not-racy this needs to be

```
auto ptr = alive.lock();
if (ptr != nullptr) {
```

That will guarantee that the pointer can't be deleted while `unpause()` is running.

Reply:
It's safe because the destruction happens on the same thread, but changing as suggested because you're right that it otherwise merits a comment explaining this, and it's easier to just make it not matter. :)